### PR TITLE
enhancement: Add a new enhancement 'filesystem-trim'

### PR DESCRIPTION
**Pull request by shuo-wu**
#836

### CODE REVIEW

**Review by innobead — APPROVED** [review]
In general, LGTM. Just two comments.

File: enhancements/20221103-filesystem-trim.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+3. Longhorn volumes won't mark snapshots as removed hence most of the time there is no continuous unavailable snapshots during the trim. 
+   To make it more practicable, we introduce a new setting/option for the volumes, which will automatically mark the latest snapshot and its ancestors as removed and stops at the snapshot containing multiple children. 
+   There is a per-volume option that can indicate if this automatic removal is enabled, or just follows a global setting.     
+
+### User Stories
+#### Reclaim the space wasted by the removed files in a filesystem
+Before the enhancement, there is no way to reclaim the space. To shrink the volume, users have to launch a new volume with a new filesystem, and copy the existing files from the old volume filesystem to the new one, then switch to use the new volume. 
+
+After the enhancement, users can directly reclaim the space by trimming the filesystem via cmd `fstrim` or Longhorn UI. Besides, users can enable the new option so that Longhorn can automatically mark the snapshot chain as removed then trim the blocks recorded in the snapshots.
```
[the remark this answers is not in the document]
Comment (shuo-wu) [reply enhancements/20221103-filesystem-trim.md:28]: NO. In the future, we can have a new recurring job type `fstrim` to handle that.

File: enhancements/20221103-filesystem-trim.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+#### instance-manager:
+- Update the dependencies.
+- Add the corresponding proxy API for the new engine API.
+
+#### longhorn-ui:
+1. Add 2 new operations for Longhorn volume.
+   - API `updateUnmapMarkSnapChainRemoved`: 
+     - The backend accepts 3 values of the input `UnmapMarkSnapChainRemoved`: `"enable"`, `"disable"`.
+     - The UI can rename this option to `Remove Current Snapshot Chain during Filesystem Trim`, and value `""` to `follow the global setting`.
```
[the remark this answers is not in the document]
Comment (shuo-wu) [reply enhancements/20221103-filesystem-trim.md:75]: Got it!

**Review by innobead — APPROVED** [review]

**Comment by derekbit** [comment]
Trim might be time-consuming if the option `UnmapMarkSnapChainRemoved` to `true`. Is a new snapshot creatable while trimming the filesystem?

**Comment by innobead** [comment]
> Trim might be time-consuming if the option `UnmapMarkSnapChainRemoved` is `true`. Is a new snapshot creatable while trimming the filesystem?

Shouldn't snapshot-related operations (trim, snapshot, purge/prune, merge/coalesce) be exclusive? otherwise, during the trim, how can we ensure these operations? @shuo-wu

**Comment by innobead** [comment]
Is this specific for which fs mount or actually it's quite general?

**Comment by derekbit** [comment]
It is quite general, and both ext4 and xfs support the option. It can make the space reclamation transparent to users.
We can have a new ticket for checking any potential issue and performance influence (might slow down the deletion operation) when the option is added.

**Comment by shuo-wu** [comment]
I think we need to have a setting that allows users to customize the mount options including `discard`. Since with the setting we introduced, the trim requests would lead to all snapshots becoming removed. This means each file removal after adding `discard` would invalidate all snapshots.

**Comment by derekbit** [comment]
You are right...BTW, user might add the mount option `discard` in the storage class' `mountOptions`, so it will lead to this issue you mentioned. Probably need to mention this point in the trim feature doc first.

**Comment by shuo-wu** [comment]
Sure. I will mention it the doc later.

**Comment by alekc** [comment]
Is there any ETA for the release  with this feature by any chance?

**Pull request merged**